src/Compilers/Core/Portable/InternalUtilities/SemaphoreExtensions.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Roslyn.Utilities
+{
+    internal static class SemaphoreExtensions
+    {
+        public static SemaphoreDisposer DisposableWait(this Semaphore semaphore, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (cancellationToken.CanBeCanceled)
+            {
+                int signalledIndex = WaitHandle.WaitAny(new[] { semaphore, cancellationToken.WaitHandle });
+                if (signalledIndex != 0)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    throw ExceptionUtilities.Unreachable;
+                }
+            }
+            else
+            {
+                semaphore.WaitOne();
+            }
+
+            return new SemaphoreDisposer(semaphore);
+        }
+
+        public static Task<SemaphoreDisposer> DisposableWaitAsync(this Semaphore semaphore, CancellationToken cancellationToken = default(CancellationToken))
```

Comment:
@bradwilson How many tests will xUnit be attempting to execute in parallel normally? If that number is close to 1, then the form used here would not cause a problem in our current code base (the only use of `Semaphore` is for this gate).

It may be possible to implement `DisposableWaitAsync` without holding a thread, but it's non-trivial to get it right with `Semaphore`. I preferred this to the alternative to placing a blocking wait inside of an unrelated `async` method.